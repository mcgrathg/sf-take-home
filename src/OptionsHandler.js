import { Component } from 'react';
import PropTypes from 'prop-types';

import { STAGES } from './constants';

const OPTIONS = [
  'Arms',
  'Shoulders',
  'Back',
  'Cleavage',
  'Midsection',
  'Rear',
  'Legs',
];

class OptionsHandler extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
  };

  state = {
    selected: {},
    stageIndex: 0,
  };

  setSelectionsForStage = (selections = []) => {
    const { stageIndex, selected } = this.state;
    const currentStage = STAGES[stageIndex];
    const newSelections = selections.reduce(
      (acc, curr) => ({
        ...acc,
        [curr]: currentStage,
      }),
      {},
    );

    this.setState({
      selected: { ...selected, ...newSelections },
      stageIndex: stageIndex + 1,
    });
  };

  render() {
    const { children } = this.props;
    const { selected, stageIndex } = this.state;

    return children({
      selectedOptions: selected,
      stageName: STAGES[stageIndex],
      stageProps: {
        options: OPTIONS.filter(option => !selected[option]),
        onSubmit: this.setSelectionsForStage,
      },
    });
  }
}

export default OptionsHandler;
