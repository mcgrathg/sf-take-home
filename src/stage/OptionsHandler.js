import { Component } from 'react';
import PropTypes from 'prop-types';

import { STAGES } from '../config/constants';
import StageSelectionHandler from './StageSelectionHandler';

const OPTIONS = [
  'Arms',
  'Shoulders',
  'Back',
  'Cleavage',
  'Midsection',
  'Rear',
  'Legs',
];

class OptionsHandler extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
  };

  state = {
    selected: {},
    stageIndex: 0,
    stageSelections: {},
  };

  setSelectionsForStage = (selections = []) => {
    const { stageIndex, selected, stageSelections } = this.state;
    const currentStage = STAGES[stageIndex];
    const newSelections = selections.reduce(
      (acc, curr) => ({
        ...acc,
        [curr]: currentStage,
      }),
      {},
    );

    const selectionsForState = new Set(stageSelections[currentStage] || []);
    selections.forEach(selection => selectionsForState.add(selection));

    this.setState({
      stageIndex: stageIndex + 1,
      selected: { ...selected, ...newSelections },
      stageSelections: {
        ...stageSelections,
        ...{ [currentStage]: Array.from(selectionsForState) },
      },
    });
  };

  render() {
    const { children } = this.props;
    const { selected, stageIndex, stageSelections } = this.state;

    return children({
      stageName: STAGES[stageIndex],
      stageSelections,
      options: OPTIONS.filter(option => !selected[option]),
      onSubmit: this.setSelectionsForStage,
    });
  }
}

export default OptionsHandler;
