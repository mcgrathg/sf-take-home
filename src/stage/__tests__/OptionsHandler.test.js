import React from 'react';
import { render } from 'react-testing-library';
import { childrenFunc } from 'test-utils';
import OptionsHandler from '../OptionsHandler';

describe('OptionsHandler', () => {
  it('renders', () => {
    const { container } = render(
      <OptionsHandler>{childrenFunc}</OptionsHandler>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
