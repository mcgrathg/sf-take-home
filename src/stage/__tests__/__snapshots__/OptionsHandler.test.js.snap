// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsHandler renders 1`] = `
<div>
  something
</div>
`;
